agent/src/protocols/udev/discovery_impl.rs

```diff
@@ -374,6 +370,64 @@ fn filter_by_remaining_udev_filters(
                     })
                     .collect();
             }
+            Rule::subsystems => {
+                mutable_devices = mutable_devices
+                    .into_iter()
+                    .filter(|device| {
+                        let is_match = device_has_parent_with_subsystem(device, &udev_filter.value);
+                        (is_equality && is_match) || (!is_equality && !is_match)
+                    })
+                    .collect();
+            }
+            Rule::attributes => {
+                let key = udev_filter
+                    .field
+                    .clone()
+                    .into_inner()
+                    .next()
+                    .unwrap()
+                    .into_inner()
+                    .next()
+                    .unwrap()
+                    .as_str();
+                mutable_devices = mutable_devices
+                    .into_iter()
+                    .filter(|device| {
+                        let is_match =
+                            device_has_parent_with_attribute(get_parent(device), key, &value_regex);
+                        (is_equality && is_match) || (!is_equality && !is_match)
+                    })
+                    .collect();
+            }
+            Rule::drivers => {
+                mutable_devices = mutable_devices
+                    .into_iter()
+                    .filter(|device| {
+                        let is_match =
+                            device_has_parent_with_driver(get_parent(device), &value_regex);
+                        (is_equality && is_match) || (!is_equality && !is_match)
+                    })
+                    .collect();
+            }
+            Rule::kernels => {
+                mutable_devices = mutable_devices
+                    .into_iter()
+                    .filter(|device| {
+                        let is_match =
+                            device_has_parent_with_sysname(get_parent(device), &value_regex);
+                        (is_equality && is_match) || (!is_equality && !is_match)
+                    })
+                    .collect();
+            }
+            Rule::tags => {
+                mutable_devices = mutable_devices
+                    .into_iter()
+                    .filter(|device| {
+                        let is_match = device_has_parent_with_tag(get_parent(device), &value_regex);
```

Comment:
it feels like there is a lot of *almost* identical code in this file that could be refactored.  i'm not sure if it can be (Rust might make that hard), but some simple things like this could be pulled into a helper method:

```
let is_match = device_has_parent_with_sysname(get_parent(device), &value_regex);
(is_equality && is_match) || (!is_equality && !is_match)
```

Reply:
i wonder if you couldn't use a closure to get even more refactoring ... probably not with Rust:

```
fn filter_udev_prop(is_match: bool) -> bool {
(is_equality && is_match) || (!is_equality && !is_match)
}
fn filter_mutuable_devices(mutable_devices, <closure>) {
  mutable_devices.into_iter().filter(|d| filter_udev_prop(<clusure>(d))).collect()
}
...
            Rule::kernels => {
                mutable_devices = filter_mutuable_devices(mutable_devices, |device| { device_has_parent_with_sysname(get_parent(device), &value_regex) });
            }
```

Reply:
I was thinking about whether i could use closures. It seems the limiting factor is the `impl Trait` `DeviceExt` that is used for testing. For example, 
```
fn filter_mutuable_devices(mutable_devices: Vec<impl DeviceExt>, f: fn(&impl DeviceExt, &Regex) -> bool, arg: &impl DeviceExt, arg2: &Regex) {
...
}
```
Is not possible because "`impl Trait` not allowed outside of function and inherent method return types"

Reply:
i feared as much.  bummer.